Add relative timestamps to conversation list

lorenzhs opened the pull request:
Fixes #284

lorenzhs commented:
Can I get some feedback on this?

liliakai commented:
Sorry for the long turnaround on this. I still need to give it a more thorough read and testing but it looks like it'll work. I might prefer to keep the date util functions on the timestamp view if they are not used anywhere else. Also "MessageTimestampView" seems slightly innacurate following this change. I wonder if it would be cleaner to sub-class a shared timestamp view rather than switching the format after constructing the view.

lorenzhs commented:
Thanks for the input! Since the date utils aren't used anywhere else I guess it wouldn't hurt to inline them. I'll also look into your subclassing idea, the current code really isn't particularly nice. I hope to get around to this on the weekend.

lorenzhs commented:
@liliakai I split the timestamp view into two subclasses and added some more tests, during which I also found and fixed a bug handling dates a month or more ago in the same year.

One potential issue is that if someone were to attempt using `TimestampView` directly now, this would not work as the function `getRelativeTimeSpanString` is defined on the children (~virtual method). Is this an issue for you, and do you have suggestions on how I could fix it?

Do you want me to squash?

liliakai commented on js/views/timestamp_view.js:
```diff
@@ -56,4 +45,61 @@
+            } else if (timediff.minutes() > 1 ||
+                       (timediff.minutes() === 1 && timediff.seconds() >= 30)) {
+                // same as hours
+                var wrap = (timediff.seconds() >= 30 ? 1 : 0)
```
jshint complains that wrap is defined twice.

liliakai replied:
Could fix that and DRY things up a bit with a  little `this.wrap(time)` helper function.

lorenzhs replied:
I keep forgetting that JS doesn't have block scopes. I cleaned up the wrapping logic (it wasn't clear to me what a function would do there, so I moved it to the top of the function).

McLoo commented:
Could we get this i18ned? 
Not only the months and days, but also the date formats? The map tells why this is needed: 😨 http://www.theguardian.com/news/datablog/2013/dec/16/why-do-americans-write-the-month-before-the-day

lorenzhs commented:
@McLoo I agree it would be nice, but when I started on this I don't recall seeing any i18n anywhere. In any case it's a separate change and shouldn't be part of this pull request in my opinion.

Given that the month is spelled as a three-letter short form ("Jan", not "1") I don't see a big problem with the current approach to be honest. It's not like there's any confusion about which is the month and which the day.

liliakai commented:
@McLoo See also #578

liliakai commented:
e876d8f6
